src/renderer/grid_renderer.rs

```diff
@@ -209,4 +193,58 @@ impl GridRenderer {
 
         canvas.restore();
     }
+
+    fn draw_underline(
+        &self,
+        canvas: &mut Canvas,
+        style: &Arc<Style>,
+        underline_style: UnderlineStyle,
+        p1: Point,
+        p2: Point,
+    ) {
+        let mut underline_paint = self.paint.clone();
+
+        underline_paint.set_color(style.special(&self.default_style.colors).to_color());
+        underline_paint.set_stroke_width(1.);
+
+        match underline_style {
+            UnderlineStyle::Underline => {
+                underline_paint.set_path_effect(None);
+                canvas.draw_line(p1, p2, &underline_paint);
+            }
+            UnderlineStyle::UnderDouble => {
+                underline_paint.set_path_effect(None);
+                canvas.draw_line(p1, p2, &underline_paint);
+                let p1 = (p1.x, p1.y - 2 as f32);
+                let p2 = (p2.x, p2.y - 2 as f32);
+                canvas.draw_line(p1, p2, &underline_paint);
+            }
+            UnderlineStyle::UnderCurl => {
+                let p1 = (p1.x, p1.y - 2.);
+                let p2 = (p2.x, p2.y - 2.);
+                underline_paint
+                    .set_path_effect(None)
+                    .set_anti_alias(true)
+                    .set_style(skia_safe::paint::Style::Stroke);
+                let mut path = Path::default();
+                path.move_to(p1);
+                let mut i = p1.0;
+                let mut sin = -2.;
+                while i <= p2.0 {
+                    sin *= -1.;
+                    i += 4.;
+                    path.quad_to((i - 2., p1.1 + sin), (i, p1.1));
+                }
+                canvas.draw_path(&path, &underline_paint);
+            }
```

Comment:
Your concerns about HiDPI and the curl appearing too thin are rightfully:

![Screenshot of Neovide showing the undercurl and underline modifiers in comparison to normally sized characters on a HiDPI setup](https://user-images.githubusercontent.com/80128916/183672695-a08dc4d8-380d-480d-9383-5b09e2e8371b.png)

I think this could be patched up by making the line slightly thicker, doing less iterations of the loop and doing larger steps each time, though this is indeed tricky. To be honest, I don't think we'll be coming to a really aesthetic solution there.

Reply:
Yeah I was afraid of this.

To be fair: no terminal I've tested handles it any better than what we're seeing here. So, maybe we could just release this like this and then improve upon it in some way?

A quick win would be to make the line thickness a configuration option, as well as maybe the pattern length (which can be used to scale the dashes and dots as well). Trying to automatically render this the best way seems like a doomed idea, or at least much harder than the rest.

Reply:
Allowing configuration seems like the most sensible option. There's lots of cases where relying on the scale factor the system gives us is wrong.

Reply:
Another idea would be to actually scale with the font height, but round the numbers so that they stay pixel perfect. Everything should still be 1px thick at low DPIs, but it should scale nicely after that, without any aliasing shenanigans. I'll see if I can come up with a quick example of this.

Reply:
Just a note, how currently everything looks like with HiDPI

![All underline styles tried out and applied on a HiDPI display. Notably the distance is too small every time.](https://user-images.githubusercontent.com/80128916/183685643-de99f5ed-e886-4f62-98a6-e8e5afda7201.png)

Reply:
Alright, so, I worked on this a little and I'm a bit stuck.

On one hand, automatic scaling looks pretty nice: ![uvNaekl](https://user-images.githubusercontent.com/4097716/184254109-a44357ff-eaa0-41c2-b7a0-48a53f6a0502.png)
 
On the other hand... this can't actually work consistently and this seems to be a limitation of the whole rendering strategy of neovide.

The issue is that everything is drawn asynchronously through a message queue, which works under the premise everything that'll get drawn (within a window) is on its own independent region of the screen, meaning there's no conflict anywhere.

But to draw thick underlines, we need them to be drawn between two lines. They can't be confined to the line the text belongs to, or they will simply strike through it at higher thicknesses (especially the undercurl). But by removing them from the line's canvas so they don't get clipped, they get randomly overwritten by the line below, leaving only two pixels to work with.

What's needed would be to redraw the underlines alone every time the line of text below them gets drawn. It seems pretty hard to do with how the code is currently set up, unless there's an easy solution I'm missing?

Reply:
I think you're right, `draw_foreground` only ever gets called with text fragments, and not the whole current text buffer. This is yet another issue which could be handled through keeping state about the _whole_ text visible at all times, and also redrawing the line below when drawing underlines. But as of now, Neovide doesn't have such a facility.

For now, the line below "sometimes" overlapping the curl is as far as you can get without taking quite a few hoops in terms of re-writing code architecture, I fear.

Reply:
It's a little better when I:
* Draw lines from the bottom up instead of the top down when drawing everything
* Redraw the line above after any line that's individually drawn
But there are still plenty of glitches here and there. I'm hopeful they can be ironed out, but it'll all just be hacks for now.

Reply:
Alright. Then feel free to push your changes, I'll create an issue about possible ways to approach the larger needed change after merging this PR.

Reply:
Yeah that's sensible.

For now, the automatic scaling works well so long as line thickness doesn't get above 1. Above that, underlines get clipped here and there. Perhaps it would make sense to put the automatic scaling behind a configuration option for now? I'll push those changes up for review.